src/modules/bulletproofs/main_impl.h

```diff
@@ -7,6 +7,115 @@
 #ifndef _SECP256K1_MODULE_BULLETPROOFS_MAIN_
 #define _SECP256K1_MODULE_BULLETPROOFS_MAIN_
 
-/* TODO */
+#include "include/secp256k1_bulletproofs.h"
+#include "include/secp256k1_generator.h"
+#include "modules/generator/main_impl.h" /* for generator_{load, save} */
+#include "hash.h"
+#include "util.h"
+
+struct secp256k1_bulletproofs_generators {
+    size_t n;
+    /* n total generators; set n = 2*k to get G_i and H_i values for i in [1..k] */
+    secp256k1_ge* gens;
+};
+
+secp256k1_bulletproofs_generators *secp256k1_bulletproofs_generators_create(const secp256k1_context *ctx, size_t n) {
+    secp256k1_bulletproofs_generators *ret;
+    secp256k1_rfc6979_hmac_sha256 rng;
+    unsigned char seed[64];
+    size_t i;
+
+    VERIFY_CHECK(ctx != NULL);
+
+    ret = (secp256k1_bulletproofs_generators *)checked_malloc(&ctx->error_callback, sizeof(*ret));
+    if (ret == NULL) {
+        return NULL;
+    }
+    ret->gens = (secp256k1_ge*)checked_malloc(&ctx->error_callback, n * sizeof(*ret->gens));
+    if (ret->gens == NULL) {
+        free(ret);
+        return NULL;
+    }
+    ret->n = n;
+
+    secp256k1_fe_get_b32(&seed[0], &secp256k1_ge_const_g.x);
+    secp256k1_fe_get_b32(&seed[32], &secp256k1_ge_const_g.y);
+
+    secp256k1_rfc6979_hmac_sha256_initialize(&rng, seed, 64);
```

Comment:
Huh, RFC6979 is terribly slow. I think that should just be a tagged hash. Or even use a tagged hash to create a ChaCha20 seed and then use ChaCha20 to expand the seed. (I think we'll need many many generators, so this could make a difference? I mean we could hardcode them in a further PR but maybe not every implementation wants to do this.)

Reply:
In practice, at least for rangeproofs we'll have a hardcoded generator list. The use of RFC6979 is just so that people can reproduce our work to generate the NUMS points.

Agreed that chacha would be better, but I don't think it's a big deal.

Reply:
> In practice, at least for rangeproofs we'll have a hardcoded generator list.

Is this list in https://github.com/ElementsProject/secp256k1-zkp/pull/207? I can't find it, but I've only skimmed that PR. I think I don't get the entire picture of the API yet.

Reply:
No, it will be hardcoded in Elements. This library tries to be agnostic about the set of generators.